### PR TITLE
[Tests] Non duplicated votes update tests

### PR DESCRIPTION
**Pull request by dkijania**
Updated chain-libs with newest changed regarding non-duplicated votes feature. Added new test cases for positive and negative scenario. Introduced FragmentChainSender which should improve test readibility and VotesCastCounter which should help adjusting load test for new requirements

### CODE REVIEW

File: testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+    pub fn new(vote_plans: Vec<(VotePlanId, u8)>) -> Self {
+        Self {
+            register: vote_plans
+                .iter()
+                .map(|(id, limit)| VotePlanVoteCastPosition::new(id.clone(), *limit))
+                .collect(),
+        }
+    }
+
+    pub fn has_next(&self) -> bool {
```
Comment (Zeegomo) [inline testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs:72]: nit: this sounds like I'm reading a java iterator implementation

File: testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+    }
+
+    pub fn available_to_send(&self) -> usize {
+        (self.limit - self.current).into()
+    }
+
+    pub fn advance_single_unsafe(&mut self) -> Result<VotesToCast, Error> {
+        self.advance_batch_unsafe(1)
+    }
+    pub fn advance_batch_unsafe(
```
Comment (Zeegomo) [inline testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs:174]: Nit: I think unsafe usually conveys a particular concept in rust, which is not what it means here

File: testing/jormungandr-testing-utils/src/testing/fragments/generator.rs [hunk 3]
```diff
@@ -206,16 +233,31 @@ impl<'a, S: SyncNode + Send> FragmentGenerator<'a, S> {
...
-                0,
-                &Choice::new(1),
-                &self.node,
-            ),
+            8 => {
+                let vote_cast_register = self
+                    .vote_cast_register
+                    .as_mut()
+                    .expect("please run 'prepare' method before running load");
+                let wallet_idx = 0;
```
Comment (Zeegomo) [inline testing/jormungandr-testing-utils/src/testing/fragments/generator.rs:241]: Why always with wallet 0?

**Review by Zeegomo — COMMENTED** [review]
Maybe I've missed it, but did you plan to update the `VoteCastsGenerator` in `testing/jormungandr-testing-utils/src/testing/fragments/load/vote_casts_generator.rs` as well?

File: testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs (hunk 1 above)
Comment (dkijania) [reply testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs:72]: I guess it is not good :) removing then

File: testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs (hunk 2 above)
Comment (dkijania) [reply testing/jormungandr-testing-utils/src/testing/vit/vote_cast_register.rs:174]: renaming

File: testing/jormungandr-testing-utils/src/testing/fragments/generator.rs (hunk 3 above)
Comment (dkijania) [reply testing/jormungandr-testing-utils/src/testing/fragments/generator.rs:241]: added comment

**Comment by mzabaluev** [comment]
Needs to be rebased, also the lints have been fixed in master.

**Pull request merged**